### ISSUE TITLE
Make Serializer.validate generic?

### ISSUE THREAD

**Issue by jaredlockhart** [typeddjango/djangorestframework-stubs#82]
Serializer.validate is currently untyped:

https://github.com/typeddjango/djangorestframework-stubs/blob/master/rest_framework-stubs/serializers.pyi#L124-L125

Is it possible to do something like

```py
T = TypeVar("T"):
class Serializer(BaseSerializer):
    def validate(self, attrs: T) -> T:
        return attrs
```

I'm experimenting with using [TypedDict](https://docs.python.org/3/library/typing.html#typing.TypedDict) to declare explicit interfaces for the data that a serializer processes, and I can annotate an overridden method like so:

```py
class MySerializer(serializers.Serializer):
    name = serializers.CharField()

    class MySerializerData(TypedDict):
        name: str

    def validate(self, data: MySerializerData) -> MySerializerData:
        reveal_type(data["name"])  # str
        validated_data = super().validate(data) # Call to untyped function "validate" in typed contextmypy(error)
```

I'm just learning how Python typing works so please forgive my lack of knowledge!

Thanks!

**Comment by jaredlockhart** [comment]
Alternatively it could be typed with `Any` to be more consistent with the other stubs:

```py
    def validate(self, attrs: Any) -> Any:
        return attrs
```

**Comment by sobolevn** [comment]
Let's make it generic!

**Comment by Goldziher** [comment]
Its neither genetic nor any, its OrderedDict

**Comment by Goldziher** [comment]
I think this issue can be closed.

**Comment by sobolevn** [comment]
But, it is still not *very* generic. Currently we only type as an `OrderedDict`. Which is not bad, but not ideal.
We know the schema type. And we can proably infer the correct type somehow.

**Comment by Goldziher** [comment]
I don't think this is possible using typing only TBH, we will need to do this using the plugin. The validated data is populated with the output of fields.

**Comment by sobolevn** [comment]
Yes, I agree. Looks like a task for the plugin.
